### ISSUE TITLE
Enhancement of the Brouter-interface

### ISSUE THREAD

**Issue by EssBee59** [osmandapp/OsmAnd#18865]
### Describe the idea (required)

Hello Osmand team
I liked to suggest an enhancement when using the Brouter for the route calculation:

==> Currently situation:
The profile to be used in Brouter is more or less defined in the brouter itself...
(for a biker, only 2 bike-profiles can be adressed by the Osmand user, with a very complex logic)

==> It would be very nice to have a new solution:
-Since Osmand version 3, many profiles can be defined in Osmand, the user can now easily switch between these profiles
-The interface to the Brouter-app now supports a new parameter "profile"

By combining the 2 features it becomes now possible for the Osmand user to define a profile in Osmand for each bike type he is using (race, trekking, MTB..) and to select dynamically the prefered bike-type.
The change below is well running for me (I currently extract the brouter-profile name from the defined Osmand-Profile names, of course a new GUI field in the profile-creation dialog should be much better)

### Tell us about the expected behaviour (required)

See description above.
I expect more flexibility (any profile defined in Brouter-app can be addressed) in a much better comprehensive way.
(for unexperienced users it is dificult to understand the current way to select/change a profile in brouter)  

### Tell us about alternatives you've considered (required)

I think, the "profile" parameter of the interface is a good solution for an enhancement.
I do not see a better  way.

### Context (optional)

I hope, the need to be able to use different profiles really exist by a lot of bikers... (not only for me and my frends)
Regards

**Comment by EssBee59** [comment]
[Osmand_PR2.txt](https://github.com/osmandapp/OsmAnd/files/13828960/Osmand_PR2.txt)

**Comment by vshcherb** [comment]
```
		Bundle bpars = new Bundle();
		bpars.putDoubleArray("lats", lats);
		bpars.putDoubleArray("lons", lons);
		bpars.putDoubleArray("nogoLats", nogoLats);
		bpars.putDoubleArray("nogoLons", nogoLons);
		bpars.putDoubleArray("nogoRadi", nogoRadi);
		bpars.putString("fast", params.fast ? "1" : "0");
		bpars.putString("v", mode);
		bpars.putString("trackFormat", "gpx");
		bpars.putString("turnInstructionFormat", "osmand");
		bpars.putString("acceptCompressedResult", "true");

// NEW
//  Extention: Send as parameter the profile-name that should be used in the Brouter for this routing-request!
//  Goal: 
//     - Till now the profile used by the brouter for a routing is more or less set in the brouter-app itself 
//          (Osmand-user can select 1 of the 6 in brouter preselected profiles with the parameters "fast"-->[0|1] and "v"-->[motorcar|bicycle|foot])
//          this is functioning, but the flexibility is very limited and the user must remain the complex mapping   
//     - Since years Osmand supports many profiles (custom-profiles). 
//     - The Brouter-app now supports in the IPC interface a new parameter "profile":
         If delivered, it provides the name of the profile to be used in brouter-app for the curent request 
//     - Using both features together it is now possible to define several profiles in Osmand that use different profiles in Brouter-app  
//          - More possibilities (every profile installed in Brouter can be addressed)
//          - Better overview     (the name of the brouter profile to be used is defined/visible in Osmand)
//          - No need to change the Brouter-app config each time you change the profile to be used (comfortable) 
//
//  ===> EXAMPLE: the "simple" change below checks the used custom-profile-name in Osmand:
//          - If the name looks like "Brouter[xxxxxx]", "xxxxxx" is considered to be the profile to be used in Brouter-app 
//                 Rule: Osmand profile_name must beginn with "Brouter", and the Brouter_Profile_Name is delimited  with brackets [...]
//                (Brouter tries to use a file named "xxxxxx.brf" in ../brouter/profile2, reports an error if not found)
//          - If the name do not like as above, the old interface is used (profile is not set)
//
//          - The Osmand user should create as much custom-profiles as he is using differents profiles in brouter
//                - Of course for each profile "navigation type" must be "BRouter (offline)"
//                - "Base Profile" and "fast" Y/N" are ignored)
//                - Examples for Osmand-Profile-Name:                              
//                    Brouter[fastbike]				(standard brouter profile for fastbike)  
//                    Brouter[fastbike-lowtraffic]		(standard brouter profile for fastbike-lowtraffic)      
//                    Brouter[racebike-verylowtraffic]		(self developped profile)
//                    Brouter[trekking]		 		(standard brouter profile for trekking)   
//                    Brouter[MTB-hard]				(self installed brouter profile)
// Note: a list of the valid brouter-profile-names can be get manually: ==> Start the "Brouter-app" AND and click on "Brouter APP"
//  

// of course, a full implementation (with a new GUI field in Osmand) would be better as the used naming convention in the simple version above 

// get "remote profile name" if format given    
 	String osmand_Profile_Name = params.mode.getUserProfileName();
     if ( osmand_Profile_Name.indexOf("Brouter") == 0)
   { 
     if ( osmand_Profile_Name .indexOf("[") != -1 && osmand_Profile_Name .indexOf("]") != -1)
   {
     String  brouter_Profile_Name = osmand_Profile_Name.substring(osmand_Profile_Name .indexOf("[") + 1, osmand_Profile_Name .indexOf("]"));

     log.info (" BROUTER_PROFILE_NAME = " + brouter_Profile_Name );
     if (brouter_Profile_Name.length() > 0)
   {
 
//  set the profile-name in the new parameter "profile" to transmit the profile-name to the brouter
          bpars.putString("profile", brouter_Profile_Name );
     
   }}}

//  end of extention....
```

**Comment by vshcherb** [comment]
OsmAnd can send as profile current profile name (lowercase) and b-router will use it. 
P.S. Could you send actual pull request to OsmAnd?
We would need to document it as well.

**Comment by EssBee59** [comment]
> OsmAnd can send as profile current profile name (lowercase) and b-router will use it.
> P.S. Could you send actual pull request to OsmAnd?
> We would need to document it as well.

Hello vshcherb.
Thank a lot for your answer and your interest for an enhancement!

You suggest to send to the Brouter-app the name of the current profile in Osmand (without the name convention + extraction I was using in the test version) ?
This is of course possible and very easy, we could prepare a PR for this....

but there were some disadvantages associated with this "simple" solution:
-old installations were not longer running (as the name of the profile would always be sent, and of course not known as a profile-name in Brouter-app) 
-the list of the profiles in Osmand would become strange / difficult to understand

 
So I would prefer something like:
- create a new GUI field in the osmand configurator to explicitly get the name of the Brouter-profile to be used
(field can be empty)
OR
- Use a naming convention (of any kind) in the osmand profile name. (See my example)
This help to better understand the profile list and to deliver an empty name by wisch.

If the GUI fieldd remain empty OR  from the naming convention the name remains empty, THEN DO NOT SEND the new parameter (compatibilty of old installations)

I hope my english is understandable and prefer to wait on a decision of the Osmand team before sending a PR
Regards

**Comment by vshcherb** [comment]
Creating UI is a long story cause UI can't be created out of nothing it needs to be in the context of BRouter configuration. As I see we don't have much specific UI for B-Router so it might be a longer story then

**Comment by afischerdev** [comment]
Added a PR for this discussion.
In my eyes it is an easy solution to get a profile name defined. But it has also all the possibilities to add a wrong name BRouter don't find in profile folder.
A more safety version would be e.g. to give the user a file dialog to have a visit in the BRouter profile folder and select a pretty profile. At this time only the profile name is needed - no other file handling have to be done.

**Comment by EssBee59** [comment]
@vshcherb 
We prepared carrefully the PR 18940 for the "simple" solution.
Plaese commit it, so we can veriy the results again with the daily build

Regards

**Comment by vshcherb** [comment]
Thanks guys, please let me know if I can close the issue

**Comment by EssBee59** [comment]
Hello,vshcherb,

I could test with the daily build Osmand+
-regression tests using the old mapping ok
-test of the new mapping ok

Yes, you can close the issue.

Here (if you are interested or like to test) a draft for the new documentation that we intend to deploy as soon the change becomes standard (with next Osmand release)

=========================
NEW for Osmand users:
=====================
From Osmand 4.7.X upwards a new mapping-mode is supported:
Since Osmand version 3, many profiles can be defined in Osmand and the user can easily switch between these profiles.
This allow now when using the service-interface to address different brouter-profiles in a more flexible and better comprehensive way.


- If in Osmand a profile has "BRouter" defined as navigation service
- AND the profile-name looks like "Brouter[mysting]

==> then the profile "mystring" will be used in the Brouter-app!
(this new mapping replaces in that case the basic mapping defined above and based on the file "serviceconfig.dat)

Examples:
Osmand-profile name	Brouter-app
=================================
Brouter[trekking]	"trekking" profile will be used (file trekking.brf)
Brouter[racebike]	"racebike" profile will be used (file racebike.brf)
....

Remark:
Currently Osmand do not check the defined name (case sensitiv) for the Brouter-profile (mystring).
If no profile is found, the routing will fail with "Could not calculate route.."!

**Comment by EugeneZmeuk** [comment]
DOCS:
- [ ] need to rewrite [the article](https://osmand.net/docs/user/navigation/routing/brouter), release OsmAnd 4.7.
   * Brouter[nav.profile]

**Comment by EssBee59** [comment]
Hello,

Yes, we plan exactly to change our documentation ==>
in "https://brouter.de/brouter/readme.txt" the draft above has to be added in the chapter "Routing via the service interface"

As you are using a link to our documentation (BRouter readme.txt), I think no change is necessary in Osmand?

Regards

**Comment by EssBee59** [comment]
> The BRouter documentation is found [here](https://github.com/abrensch/brouter/tree/master/docs).

of course no change direct on the server!
(I just wanted here to explain our Osmand friends what will change at the end)

**Comment by EssBee59** [comment]
Hello,

our documentation about the new mapping is now up to date.
The link:

https://github.com/afischerdev/brouter/blob/docs/docs/users/osmand.md

Regards
EssBee

**Comment by EugeneZmeuk** [comment]
Hello
@EssBee59, Thanks

Eugene

**Comment by afischerdev** [comment]
Sorry I'm a little bit late.
The correct link to the doc is 
[https://github.com/abrensch/brouter/blob/master/docs/users/osmand.md](https://github.com/abrensch/brouter/blob/master/docs/users/osmand.md)
 It was commited yesterday.
